### PR TITLE
Ignore SCSS/LESS variables when checking property-related rules

### PR DESCRIPTION
**Pull request by borodean**
Implements what's proposed in https://github.com/stylelint/stylelint/issues/518

@davidtheclark @jeddy3 am I missing something?

### CODE REVIEW

**Comment by TrySound** [comment]
Will such things be added for every new non-standard syntax? Caz, I want to start my variables names with curly brackets, or maybe unclosed quotes.

**Comment by borodean** [comment]
@TrySound I guess then it need to be a configurable option. @davidtheclark @jeddy3 what do you think?

**Comment by borodean** [comment]
The other option could be that parser developers mark things like variables translated to properties or line comments translated to block ones with a special key.

**Comment by borodean** [comment]
...Or we consider that people using some rare and very specific syntaxes (like variable names starting with curly brackets) should disable this rule and use a custom one.

**Comment by borodean** [comment]
...Or we can check if the property name if a valid CSS property, otherwise consider it is not. Then any custom syntaxes would be filtered out.

**Comment by borodean** [comment]
I've replaced this check
```js
if (prop[0] === "$" || prop[0] === "@") { return }
```

With this one:

```js
const RE_VALID_CSS_PROPERTY = /[a-z-]/
// ...
if (RE_VALID_CSS_PROPERTY.test(prop)) { return }
```

This doesn’t strictly follow the CSS spec:

image
http://www.w3.org/TR/css-syntax-3/#typedef-ident-token

But it pretty much covers all the existing properties, including vendor ones.

**Comment by davidtheclark** [comment]
In the regexp: Why not follow the spec more closely and also allow capital letters, underscores, and non-ASCII characters?

Personally, though, I'm not too worried about the original (faster) naive approach, because I don't think we'll have a lot of requests for dealing with weird custom syntaxes. When people come asking us to accommodate their special variables starting with curly brackets, we could reconsider. Until then, I'd be fine with the simple way. @jeddy3 any thoughts?

File: src/rules/root-no-standard-properties/index.js [hunk 1]
```diff
@@ -4,6 +4,8 @@ import {
   validateOptions,
 } from "../../utils"
 
+const RE_VALID_CSS_PROPERTY = /^[a-z-]+$/
```
Comment (davidtheclark) [inline src/rules/root-no-standard-properties/index.js]: If we have a standard, reused way of checking whether a property is a valid CSS property, we could put it in `utils/` with the other reused functions like that.

**Comment by borodean** [comment]
@davidtheclark,
> In the regexp: Why not follow the spec more closely and also allow capital letters, underscores, and non-ASCII characters?

Throught 20 years of CSS history the only exception to that RegExp were IE6/7 hacks like `*color` and `_color` and which are irrelevant nowadays. There is also no evidence that someone’s going to break this dashed-case tradition. And if someone does - this RegExp could be fixed.

Also, PostCSS by its nature parses CSS very loosely, so I see this loose check as a rather consistent one to the whole ecosystem.

I think its a good place to KISS.

File: src/rules/root-no-standard-properties/index.js (hunk 1 above)
Comment (borodean) [inline src/rules/root-no-standard-properties/index.js]: I’ll wait until you decide whether to check for `@`/`$` or the RegExp and than either fix this or rollback the RegExp.

**Comment by davidtheclark** [comment]
> Throught 20 years of CSS history the only exception to that RegExp were IE6/7 hacks like *color and _color and which are irrelevant nowadays. There is also no evidence that someone’s going to break this dashed-case tradition. And if someone does - this RegExp could be fixed.

Sounds like solid reasoning to me.

**Comment by davidtheclark** [comment]
@borodean What are your thoughts on the regexp vs naive `@|$` approach?

**Comment by borodean** [comment]
I've tested how fast those are:
image
image
http://jsperf.com/stylelint-520-indexof-vs-regexp/6

**Comment by borodean** [comment]
Node 4.2.2:

```
checkByFirstChar(property) x 76,710,943 ops/sec ±1.67% (92 runs sampled)
checkByFirstChar(scssVariable) x 75,718,663 ops/sec ±0.34% (89 runs sampled)
checkByFirstChar(lessVariable) x 68,338,943 ops/sec ±0.49% (92 runs sampled)
checkByExpression(property) x 18,314,081 ops/sec ±0.59% (101 runs sampled)
checkByExpression(scssVariable) x 27,982,717 ops/sec ±0.77% (101 runs sampled)
checkByExpression(lessVariable) x 27,369,798 ops/sec ±1.23% (94 runs sampled)
Fastest is checkByFirstChar(scssVariable)
```

**Comment by davidtheclark** [comment]
That seals it for me. Let's go with the checkByFirstChar approach and just change it down the line if necessary.

**Comment by borodean** [comment]
@davidtheclark I've extracted the validation function and rolled back the RegExp check if favor of the naive one.

**Comment by davidtheclark** [comment]
Merged in c9d1967. I made the slight change of making the utility `cssPropertyIsVariable` and added a check for `--custom-properties`, figuring that in the cases you've addressed we should ignore those as well. Thanks for the contributions @borodean!

**Comment by jeddy3** [comment]
> Thanks for the contributions @borodean!

Seconded. This is great thanks!

**Comment by borodean** [comment]
@davidtheclark @jeddy3 always a pleasure!

**Pull request closed without merging**